libraries/cbor/src/macros.rs

```diff
@@ -12,6 +12,142 @@
+///
+/// **Important:** Keys passed to the `destructure_cbor_map!` macro **must be sorted** in increasing
+/// order. If not, the algorithm can yield incorrect results, such a assigning `None` to a variable
+/// even if the corresponding key existed in the map. **No runtime checks** are made for this in the
+/// `destructure_cbor_map!` macro, in order to avoid overhead at runtime. However, assertions that
+/// keys are sorted are added in `cfg(test)` mode, so that unit tests can verify ahead of time that
+/// the keys are indeed sorted. This macro is therefore **not suitable for dynamic keys** that can
+/// change at runtime.
+///
+/// Semantically, provided that the keys are sorted as specified above, the following two snippets
+/// of code are equivalent, but the `destructure_cbor_map!` version is more optimized, as it doesn't
+/// re-balance the `BTreeMap` for each key, contrary to the `BTreeMap::remove` operations.
+///
+/// ```rust
+/// # extern crate alloc;
+/// # #[macro_use]
+/// # extern crate cbor;
+/// #
+/// # fn main() {
+/// #     let map = alloc::collections::BTreeMap::new();
+/// destructure_cbor_map! {
+///     let {
+///         1 => x,
+///         "key" => y,
+///     } = map;
+/// };
+/// # }
+/// ```
+///
+/// ```rust
+/// # extern crate alloc;
+/// # #[macro_use]
+/// # extern crate cbor;
+/// #
+/// # fn main() {
+/// #     let mut map = alloc::collections::BTreeMap::<cbor::KeyType, _>::new();
+/// use cbor::values::IntoCborKey;
+/// let x: Option<cbor::Value> = map.remove(&1.into_cbor_key());
+/// let y: Option<cbor::Value> = map.remove(&"key".into_cbor_key());
+/// # }
+/// ```
+#[macro_export]
+macro_rules! destructure_cbor_map {
+    ( let { $( $key:expr => $variable:ident, )+ } = $map:expr; ) => {
+        // A pre-requisite for this algorithm to work is that the keys to extract from the map are
+        // sorted - the behavior is unspecified if the keys are not sorted.
+        // Therefore, in test mode we add assertions that the keys are indeed sorted.
+        #[cfg(test)]
+        assert_sorted_keys!($( $key, )+);
+
+        use $crate::values::{IntoCborKey, Value};
+        use $crate::macros::destructure_cbor_map_peek_value;
+
+        // This algorithm first converts the map into a peekable iterator - whose items are sorted
+        // in strictly increasing order of keys. Then, the repeated calls to the "peek value"
+        // helper function will consume this iterator and yield values (or `None`) when reaching
+        // the keys to extract.
+        //
+        // This is where the pre-requisite that keys to extract are sorted is important: the
+        // algorithm does a single linear scan over the iterator and therefore keys to extract have
+        // to come in the same order (i.e. sorted).
+        let mut it = $map.into_iter().peekable();
+        $(
+        let $variable: Option<Value> = destructure_cbor_map_peek_value(&mut it, $key.into_cbor_key());
```

Comment:
`$key.into_cbor_key()` is constant. The compiler probably optimizes it (common subexpression elimination) but it's not guaranteed if it cannot detect that `into_cbor_key` is pure. If you care about performance, it would be useful to factor it out:

```
let key = $key.into_cbor_key();
$(
let $variable ... (&mut it, &key);
)+
```

You also need to change the signature of the helper function to take the key by reference since `KeyType` is not `Copy`.

Reply:
Sorry, I mixed it up :-) $key is *not* constant. So we have to compute it each time.

Reply:
Yes, it's `$($key)+`